### PR TITLE
Retrieve file statistics

### PR DESCRIPTION
**Pull request by davidstyers**
Supersedes  #14 This pull request aims to add enhanced functionality as requested in issue #13. I chose to solve the issue by going the second route I mention, by adding the functionality directly on the file handlers themselves and not the FileItem. This was taking the path of least resistance as there would be no need to subclass FileItem for each handler to implement the functionality.

Currently, the only handler that I was not able to implement all of the functionality in was the S3 handler. Accessed time and created time are not currently possible to implement. Other than that, all handlers include the enhancements.

### CODE REVIEW

File: filestorage/handler_base.py [hunk 1]
```diff
@@ -226,6 +283,82 @@ async def _async_exists(self, item: FileItem) -> bool:
         """Determine if the given filename exists in the storage container."""
         pass
 
+    async def async_size(self, filename: str) -> int:
```
Comment (ianepperson) [inline filestorage/handler_base.py:286]: I think these should continue to follow the `get_` pattern: `get_size()` and `async_get_size()`. "Size" by itself can be a verb in English - and would be a funky command to implement! ("Make my file this size!" ... truncates file to that size... )

File: filestorage/handler_base.py [hunk 2]
```diff
@@ -226,6 +283,82 @@ async def _async_exists(self, item: FileItem) -> bool:
...
+        """Retrieve file size for file in storage container given filename.
+        Returns the size, in bytes of the file.
+        """
+        item = self.get_item(filename)
+        return await self._async_size(item)
+
+    def _size(self, item: FileItem) -> int:
+        if not self.allow_sync_methods:
+            raise RuntimeError('Sync exists method not allowed')
```
Comment (ianepperson) [inline filestorage/handler_base.py:295]: Please update the error text (copy/paste error).

File: filestorage/handler_base.py [hunk 3]
```diff
@@ -226,6 +283,82 @@ async def _async_exists(self, item: FileItem) -> bool:
...
+    async def async_get_accessed_time(self, filename: str) -> datetime:
+        """Retrieve time of last access for file in storage container given filename.
+        Returns the time of last access, in number of seconds since the epoch.
+        """
+        item = self.get_item(filename)
+        return await self._async_get_accessed_time(item)
+
+    def _get_accessed_time(self, item: FileItem) -> datetime:
+        if not self.allow_sync_methods:
+            raise RuntimeError('Sync exists method not allowed')
```
Comment (ianepperson) [inline filestorage/handler_base.py:314]: Please update the error text (copy/paste error).

File: filestorage/handler_base.py [hunk 4]
```diff
@@ -226,6 +283,82 @@ async def _async_exists(self, item: FileItem) -> bool:
...
+    async def async_get_created_time(self, filename: str) -> datetime:
+        """Retrieve creation time for file in storage container given filename.
+        Returns the creation time, in number of seconds since the epoch.
+        """
+        item = self.get_item(filename)
+        return await self._async_get_created_time(item)
+
+    def _get_created_time(self, item: FileItem) -> datetime:
+        if not self.allow_sync_methods:
+            raise RuntimeError('Sync exists method not allowed')
```
Comment (ianepperson) [inline filestorage/handler_base.py:333]: Please update the error text (copy/paste error).

File: filestorage/handler_base.py [hunk 5]
```diff
@@ -226,6 +283,82 @@ async def _async_exists(self, item: FileItem) -> bool:
...
+    async def async_get_modified_time(self, filename: str) -> datetime:
+        """Retrieve time of last modification for file in storage container given filename.
+        Returns the time of last modification, in number of seconds since the epoch.
+        """
+        item = self.get_item(filename)
+        return await self._async_get_modified_time(item)
+
+    def _get_modified_time(self, item: FileItem) -> datetime:
+        if not self.allow_sync_methods:
+            raise RuntimeError('Sync exists method not allowed')
```
Comment (ianepperson) [inline filestorage/handler_base.py:352]: Please update the error text (copy/paste error).

File: filestorage/handler_base.py [hunk 6]
```diff
@@ -346,6 +479,54 @@ async def _async_exists(self, item: FileItem) -> bool:
         item = self._get_subfolder_file_item(item)
         return await self._store.async_handler._async_exists(item)
 
+    # Pass through any size methods
+
+    def _size(self, item: FileItem) -> int:
```
Comment (ianepperson) [inline filestorage/handler_base.py:484]: If we go with the `get_` prefix, needs to be updated here as well.

File: filestorage/handlers/dummy.py [hunk 7]
```diff
@@ -10,7 +11,9 @@ def __init__(self, **kwargs):
         super().__init__(**kwargs)
         # Store files where the key is the url path and the value is
         # the bytes that were stored to the file.
-        self.files: Dict[str, bytes] = {}
+        self.files: Dict[
+            str, Dict[str, bytes, str, datetime, str, datetime, str, datetime]
```
Comment (ianepperson) [inline filestorage/handlers/dummy.py:15]: Getting complex enough that a named tuple might be a good idea. I think it's fine as-is though.

**Review by ianepperson — CHANGES_REQUESTED** [review]
Looking pretty good and thorough! Just a few text changes and terminology suggestions.

**Comment by ianepperson** [comment]
Oh and don't forget to deal with the reference frame!
https://kosgd.medium.com/datetime-changes-in-python-4-0-474045337b99
(Please be sure to read the final update on the article prior to implementation)

**Comment by davidstyers** [comment]
Congratulations on the new job! I will look over your review as well as the article and make the requisite adjustments.

File: filestorage/handler_base.py (hunk 1 above)
Comment (davidstyers) [reply filestorage/handler_base.py]: I agree, my changes should reflect this new verbiage.

File: filestorage/handlers/dummy.py (hunk 7 above)
Comment (davidstyers) [reply filestorage/handlers/dummy.py]: I updated this to use a named tuple, a lot cleaner.

**Comment by davidstyers** [comment]
> Oh and don't forget to deal with the reference frame!
> https://kosgd.medium.com/datetime-changes-in-python-4-0-474045337b99
> (Please be sure to read the final update on the article prior to implementation)

Happy late April fools to you as well; you got me. I was going through the article thinking to myself "man, this is a really niche issue to be put into the standard library, not mentioning the awkward code it produces" I should have known better!

**Review by ianepperson — APPROVED** [review]

**Comment by ianepperson** [comment]
> Happy late April fools to you as well

That article got me as well, and I couldn't resist passing it along. If you check out the linked discussion, someone had asked for a ridiculous precision for the datetime object - so much so that relativistic effects would then start to have an effect and "now" has less meaning.

Thank you for this contribution!

**Pull request merged**